backend/dataall/modules/redshift_datasets_shares/aws/redshift_data.py

```diff
@@ -0,0 +1,408 @@
+            allowed_error_message = f'ERROR: Schema {schema} is already added to the datashare {datashare}'
+            error_message = e.args[0]
+            if error_message == allowed_error_message:
+                log.info(f'{schema=} is already present in {datashare=}')
+            else:
+                log.error(f'Adding {schema=} to {datashare=} failed due to: {e}')
+                raise e
+
+    def check_schema_in_datashare(self, datashare: str, schema: str) -> bool:
+        """Check that schema exists in datashare by describing datashare"""
+        try:
+            log.info(f'Checking {schema=} in {datashare=}...')
+            sql_statement = f'DESC DATASHARE {RedshiftShareDataClient.parsed_name(datashare)};'
+            records = self._execute_statement_return_records(sql=sql_statement)
+            schemas_in_datashare = [
+                [d for d in record][5]['stringValue']
+                for record in records
+                if [d for d in record][4]['stringValue'] in ['schema']
+            ]
+            log.info(f'Found {schemas_in_datashare=}')
+            return schema in schemas_in_datashare
+        except Exception as e:
+            log.error(f'Checking of {schema=} in {datashare=} failed due to: {e}')
+            return False
+
+    def add_table_to_datashare(self, datashare: str, schema: str, table_name: str):
+        """Add table to datashare if not already added"""
+        try:
+            log.info(f'Adding table {table_name=} to {datashare=}...')
+            table = RedshiftShareDataClient.parsed_object_names(self.database, schema, table_name)
+            sql_statement = f'ALTER DATASHARE {RedshiftShareDataClient.parsed_name(datashare)} ADD TABLE {table};'
+            self._execute_statement(sql_statement)
+        except Exception as e:
+            allowed_error_message = f'ERROR: Relation {table_name} is already added to the datashare {datashare}'
+            error_message = e.args[0]
+            if error_message == allowed_error_message:
+                log.info(f'Table {table_name} is already present in the {datashare=}')
+            else:
+                log.error(f'Adding {table_name=} of {datashare=} failed due to: {e}')
+                raise e
+
+    def check_table_in_datashare(self, datashare: str, table_name: str) -> bool:
+        """Check that table exists in datashare by describing datashare"""
+        try:
+            log.info(f'Checking {table_name=} in {datashare=}...')
+            sql_statement = f'DESC DATASHARE {RedshiftShareDataClient.parsed_name(datashare)};'
+            records = self._execute_statement_return_records(sql=sql_statement)
+            tables_in_datashare = [
+                [d for d in record][5]['stringValue'].split('.')[-1]  # "schemaname.tablename"
+                for record in records
+                if [d for d in record][4]['stringValue'] in ['table']
+            ]
+            log.info(f'Found {tables_in_datashare=}')
```

Comment:
can you give an example of records to better understand this list comprehension we are doing? Do we need to use index based retrieval?

Reply:
It is a bit different from typical API calls, The API call is the [get_statement_result](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/redshift-data/client/get_statement_result.html) which returns:
```
{
    'ColumnMetadata': ....,
    'NextToken': 'string',
    'Records': [
        [
            {
                'blobValue': b'bytes',
                'booleanValue': True|False,
                'doubleValue': 123.0,
                'isNull': True|False,
                'longValue': 123,
                'stringValue': 'string'
            },
        ],
    ],
    'TotalNumRows': 123
}
```

Reply:
In:
```
tables_in_datashare = [
                [d for d in record][5]['stringValue'].split('.')[-1]  # "schemaname.tablename"
                for record in records
                if [d for d in record][4]['stringValue'] in ['table']
            ]
```
We are processing a response of the type:
```
records = [
    [
        {
            'longValue': 1111111111111,
        }, #producer_account
        {
            'longValue': 12345689-1940u23085u2309,
        }, #producer_namespace
        {
            'stringValue': 'OUTBOUND'
        }, #share_type
        {
            'stringValue': 'string'
        }, #share_name
        {
            'stringValue': 'TABLE'
        }, #object_type
        {
            'stringValue': 'schema1.table_name_1'
        },#object_name
        {
            'booleanValue': False
        },#include_new
    ],  #Record1
]
```

Reply:
In cases where we are doing a direct SELECT a, b, c from table, we have more control over the the response fields, but for the standard SHOW DATASHARES, DESC DATASHARE etc, the fields are already provided

Reply:
makes sense - seems like not much we can do here given API response